### PR TITLE
feat: added intro splash screen

### PR DESCRIPTION
**Pull request by Bhavnaharitsa**
Fixes: #881 
Added intro splash screen which appears when the app launches for first-time or when the users kills and reopens the app.

- [.] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [.] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [.] If you have multiple commits please combine them into one commit by squashing them.

### CODE REVIEW

File: mifospay/src/main/res/layout/activity_splash_screen.xml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        android:layout_height="wrap_content"
+        android:src="@mipmap/ic_launcher_round"
+        />
+
+    <TextView
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:textColor="@color/gray_5"
+        android:text="Mobile Wallet  4.0"
+        android:textSize="30dp"
```
Comment (jawidMuhammadi) [inline mifospay/src/main/res/layout/activity_splash_screen.xml:23]: Text size should be `sp` not `dp`

File: mifospay/src/main/res/layout/activity_splash_screen.xml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:src="@mipmap/ic_launcher_round"
+        />
+
+    <TextView
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:textColor="@color/gray_5"
+        android:text="Mobile Wallet  4.0"
```
Comment (jawidMuhammadi) [inline mifospay/src/main/res/layout/activity_splash_screen.xml:22]: It is better if we use dynamic, instead of hardcoded strings

**Comment by refactor-droidyy** [comment]
@Bhavnaharitsa hey, it would be better if we don't add more activity for the splash screen it can be implemented without adding another activity I guess.

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/SplashScreen.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import android.content.Intent;
+import android.os.Handler;
+import android.support.v7.app.AppCompatActivity;
+import android.os.Bundle;
+import android.view.View;
+
+import org.mifos.mobilewallet.mifospay.auth.ui.LoginActivity;
+
+public class SplashScreen extends AppCompatActivity {
+    public static final int splash_screen_timer = 2000;
```
Comment (jawidMuhammadi) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/SplashScreen.java:12]: Use the camel case naming style please, or as it is a final variable you can write all the letters in the capital.

File: mifospay/src/main/res/layout/activity_splash_screen.xml (hunk 1 above)
Comment (Bhavnaharitsa) [reply mifospay/src/main/res/layout/activity_splash_screen.xml:23]: Okay

**Comment by Bhavnaharitsa** [comment]
> @Bhavnaharitsa hey, it would be better if we don't add more activity for the splash screen it can be implemented without adding another activity I guess.

How come?

**Comment by rachittt** [comment]
I personally don't think there is any need for the Text "Mobile Wallet 4.0"

The logo itself would do.

**Comment by Bhavnaharitsa** [comment]
> I personally don't think there is any need for the Text "Mobile Wallet 4.0"
> The logo itself would do.

Okay!!

File: mifospay/src/main/res/layout/activity_splash_screen.xml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    android:layout_height="match_parent"
+    tools:context=".SplashScreen"
+    android:layout_gravity="center"
+    android:gravity="center"
+    android:orientation="vertical">
+
+    <ImageView
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:src="@mipmap/ic_launcher_round"
```
Comment (dubeyaayush07) [inline mifospay/src/main/res/layout/activity_splash_screen.xml:15]: It will be better to use a higher resolution version of the logo. `ic_launcher_round` is a low res version suited for the app icon

**Comment by Bhavnaharitsa** [comment]
> I personally don't think there is any need for the Text "Mobile Wallet 4.0"
> The logo itself would do.

Okay, shall change it

File: mifospay/src/main/res/layout/activity_splash_screen.xml (hunk 4 above)
Comment (Bhavnaharitsa) [reply mifospay/src/main/res/layout/activity_splash_screen.xml:15]: Which one would you suggest is a higher version image?

Comment (dubeyaayush07) [reply mifospay/src/main/res/layout/activity_splash_screen.xml:15]: You can obtain the original images from [here](https://drive.google.com/drive/folders/10NTOigKiTYPH0ONbIvdGvVZHrnosQNyy)

**Comment by devansh-299** [comment]
@Bhavnaharitsa 
You don't need to create a separate layout file as loading the layout after initialization completely defies the the very use of Splash screens . You can read about "Cold Start" to know more about this . 
You can make a drawable file and pass it as arguement to android:windowBackground attribute .

**Comment by Bhavnaharitsa** [comment]
> @Bhavnaharitsa
> You don't need to create a separate layout file as loading the layout after initialization completely defies the the very use of Splash screens . You can read about "Cold Start" to know more about this .
> You can make a drawable file and pass it as arguement to android:windowBackground attribute .

Yes, got it. Let me hear from the mentors as well before I do the change :)
Thank you for suggesting though!

**Comment by luckyman20** [comment]
I agree with @devansh-299 although I am not one of the mentors.

**Comment by Bhavnaharitsa** [comment]
> I agree with @devansh-299 although I am not one of the mentors.

Okay sir, shall do it accordingly

**Comment by jawidMuhammadi** [comment]
@Bhavnaharitsa I also agree with @devansh-299 take a look in [here](https://android.jlelse.eu/launch-screen-in-android-the-right-way-aca7e8c31f52), it might help you.

**Comment by Bhavnaharitsa** [comment]
> @Bhavnaharitsa I also agree with @devansh-299 take a look in [here](https://android.jlelse.eu/launch-screen-in-android-the-right-way-aca7e8c31f52), it might help you.

Thank you! :)

**Comment by rachittt** [comment]
@Bhavnaharitsa hey, please make the necessary changes asap.

**Comment by Bhavnaharitsa** [comment]
> @Bhavnaharitsa Try adding some simple animation to it - like fade in.

Okay sir. But, is it possible in the cold start method where we use `android:windowBackground` in styles?
I think, we need to keep the activity for it. Please enlighten :)

**Comment by Bhavnaharitsa** [comment]
ezgif com-video-to-gif
@luckyman20 This is a video where I implemented fade in effect, please have a look

**Comment by jawidMuhammadi** [comment]
> > @Bhavnaharitsa hey, it would be better if we don't add more activity for the splash screen it can be implemented without adding another activity I guess.
> 
> How come?

One of the important aims of the splash screen is to avoid showing a white screen to users until the start of an activity. So whenever we use an activity to implement it then we won't achieve the exact purpose of Splash Screen. For more details about how to implement, please visit [here](https://android.jlelse.eu/launch-screen-in-android-the-right-way-aca7e8c31f52)

**Comment by luckyman20** [comment]
I meant fade in effect on the logo itself.

**Comment by Bhavnaharitsa** [comment]
> I meant fade in effect on the logo itself.

Okay, shall put it in logo

**Comment by luckyman20** [comment]
If we don't see any update in 3 days, we will have to close it as it is blocking other contributors to work on this issue.

**Comment by Bhavnaharitsa** [comment]
> If we don't see any update in 3 days, we will have to close it as it is blocking other contributors to work on this issue.

Sure sir, will give update on it

File: mifospay/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -16,16 +16,11 @@
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
         android:theme="@style/AppTheme">
-        <activity android:name=".merchants.ui.MerchantTransferActivity"></activity>
```
Comment (rachittt) [inline mifospay/src/main/AndroidManifest.xml:19]: this change is not necessary. please remove it.

File: mifospay/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -16,16 +16,11 @@
...
         <activity
             android:name=".auth.ui.LoginActivity"
-            android:theme="@style/AppTheme"
+            android:theme="@style/SplashTheme"
             android:windowSoftInputMode="stateVisible">
-            <intent-filter>
-                <action android:name="android.intent.action.MAIN" />
-
-                <category android:name="android.intent.category.LAUNCHER" />
-            </intent-filter>
```
Comment (dubeyaayush07) [inline mifospay/src/main/AndroidManifest.xml:28]: I don't think it is wise to remove the intent filter, launcher applications use this intent-filter to start the app. Are you sure the app is working fine without the intent filter?

File: mifospay/src/main/AndroidManifest.xml (hunk 5 above)
Comment (Bhavnaharitsa) [reply mifospay/src/main/AndroidManifest.xml:19]: Okay

File: mifospay/src/main/AndroidManifest.xml [hunk 7]
```diff
@@ -39,6 +38,7 @@
         <activity
             android:name=".qr.ui.ShowQrActivity"
             android:theme="@style/AppTheme" />
+
```
Comment (rachittt) [inline mifospay/src/main/AndroidManifest.xml:41]: remove this unnecessary change.

File: mifospay/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -48,6 +48,7 @@
                 android:name="android.support.FILE_PROVIDER_PATHS"
                 android:resource="@xml/provider_paths" />
         </provider>
+
```
Comment (rachittt) [inline mifospay/src/main/AndroidManifest.xml:51]: remove this unnecessary change.

**Comment by Bhavnaharitsa** [comment]
@rachittt Here you go,
ezgif com-video-to-gif

**Comment by Bhavnaharitsa** [comment]
@luckyman20 Please review!!

**Comment by luckyman20** [comment]
@Bhavnaharitsa I still cannot see logo being fading out.

**Comment by Bhavnaharitsa** [comment]
@luckyman20 I am not sure how to add animation when I do splash screen by adding it in themes. But, I suggest we open a new Activity for it because we might need it for further features(example: like loading of datas, may be some permissions and also in figuring whether the user needs to register or login) and I guess it doesn't make the app heavy-weight or anything as such and hence we can consider opening it.Just a suggestion, sir. In Activity, it is clearly possible to add the animation but if it is possible, could you help me with this?
Thank you

**Comment by HaripriyaB** [comment]
Should the splash screen be like this??
@luckyman20 
mifospaysplash

**Comment by luckyman20** [comment]
@HaripriyaB Yes.

**Comment by HaripriyaB** [comment]
> 
> 
> @HaripriyaB Yes.

okay so if you permit then I can issue a PR for this issue(if I'm allowed).

**Comment by Bhavnaharitsa** [comment]
@HaripriyaB Have you used the activity approach or by declaring it in themes? 
As you can see, there is a pr already clearly. After the debate on finalising which method to use, I shall make the changes. :)

**Comment by HaripriyaB** [comment]
> 
> 
> @HaripriyaB Have you used the activity approach or by declaring it in themes?
> As you can see, there is a pr already clearly. After the debate on finalising which method to use, I shall make the changes. :)
 For implementing the fading out feature I did using only themes.

**Comment by HaripriyaB** [comment]
> 
> 
> @HaripriyaB Have you used the activity approach or by declaring it in themes?
> As you can see, there is a pr already clearly. After the debate on finalising which method to use, I shall make the changes. :)

yes, I've used only themes for fading out feature.

**Comment by Bhavnaharitsa** [comment]
> @luckyman20 I am not sure how to add animation when I do splash screen by adding it in themes. But, I suggest we open a new Activity for it because we might need it for further features(example: like loading of datas, may be some permissions and also in figuring whether the user needs to register or login) and I guess it doesn't make the app heavy-weight or anything as such and hence we can consider opening it.Just a suggestion, sir. In Activity, it is clearly possible to add the animation but if it is possible, could you help me with this?
> Thank you

@luckyman20 Please express your view on this sir

File: mifospay/src/main/AndroidManifest.xml (hunk 5 above)
Comment (devansh-299) [reply mifospay/src/main/AndroidManifest.xml:19]: @Bhavnaharitsa please remove the unnecssary changes from this PR

Comment (Bhavnaharitsa) [reply mifospay/src/main/AndroidManifest.xml]: Done

File: mifospay/src/main/AndroidManifest.xml [hunk 9]
```diff
@@ -56,8 +55,8 @@
             android:theme="@style/AppTheme" />
         <activity
             android:name=".receipt.ui.ReceiptActivity"
-            android:theme="@style/AppTheme"
-            android:launchMode="singleTask">
+            android:launchMode="singleTask"
+            android:theme="@style/AppTheme">
```
Comment (rachittt) [inline mifospay/src/main/AndroidManifest.xml:59]: was this rearranging necessary of lines?

File: mifospay/src/main/AndroidManifest.xml [hunk 10]
```diff
@@ -71,8 +70,8 @@
         </activity>
         <activity
             android:name=".invoice.ui.InvoiceActivity"
-            android:theme="@style/AppTheme"
-            android:launchMode="singleTask">
+            android:launchMode="singleTask"
+            android:theme="@style/AppTheme">
```
Comment (rachittt) [inline mifospay/src/main/AndroidManifest.xml:74]: was this rearranging of lines necessary?

File: mifospay/src/main/AndroidManifest.xml (hunk 9 above)
Comment (Bhavnaharitsa) [reply mifospay/src/main/AndroidManifest.xml:59]: No, I didn't do it manually though. I guess it happened. But this isnt an issue right? I'll make sure it doesn't happen in my other prs

File: mifospay/src/main/AndroidManifest.xml (hunk 10 above)
Comment (Bhavnaharitsa) [reply mifospay/src/main/AndroidManifest.xml:74]: Okay, I will.

Comment (Bhavnaharitsa) [reply mifospay/src/main/AndroidManifest.xml]: done

**Comment by devansh-299** [comment]
> @luckyman20 I am not sure how to add animation when I do splash screen by adding it in themes. But, I suggest we open a new Activity for it because we might need it for further features(example: like loading of datas, may be some permissions and also in figuring whether the user needs to register or login) and I guess it doesn't make the app heavy-weight or anything as such and hence we can consider opening it.Just a suggestion, sir. In Activity, it is clearly possible to add the animation but if it is possible, could you help me with this?
> Thank you

@luckyman20 I agree with @Bhavnaharitsa . Using Splash theme method, I couldn't 
find any way to incorporate animations. 
@HaripriyaB can you please share your approach

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/auth/ui/LoginActivity.java [hunk 11]
```diff
@@ -67,6 +67,7 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         getActivityComponent().inject(this);
         setContentView(R.layout.activity_login);
```
Comment (devansh-299) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/auth/ui/LoginActivity.java:69]: @Bhavnaharitsa `activity_login` layout file won't be required after using the `SplashTheme`. You can remove this line and delete the layout file

**Comment by HaripriyaB** [comment]
I used theme approach and increased the fade out time to 1000.

**Comment by devansh-299** [comment]
@Bhavnaharitsa Your PR won't work as of your latest commit. You can follow [this](https://medium.com/@ssaurel/create-a-splash-screen-on-android-the-right-way-93d6fb444857) article for reference. You can create a splash activity and set the `SplashTheme` to that activity.

**Comment by devansh-299** [comment]
> I used theme approach and increased the fade out animation time to 1000.



> I used theme approach and increased the fade out animation time to 1000.

Can you please elaborate? How did you apply animation to the theme itself?

**Comment by Bhavnaharitsa** [comment]
@devansh-299 Yes, cause the discussion on which method to implement. Now, I don't know if we have finalised on the method we need to use. Anyway, let's hear out from @HaripriyaB about her approach.

**Comment by HaripriyaB** [comment]
I actually created a splashactivity java file and made a fade out animator with android:interpolator="@android:interpolator/accelerate_cubic" and android:duration="1000". Then I made a splash theme. and connected both with transition in splashactivity.

**Comment by Bhavnaharitsa** [comment]
> I actually created a splashactivity java file and made a fade out animator with android:interpolator="@android:interpolator/accelerate_cubic" and android:duration="1000". Then I made a splash theme. and connected both with transition in splashactivity.

So you made a new activity then and not the cold start method?

**Comment by HaripriyaB** [comment]
> 
> 
> > I actually created a splashactivity java file and made a fade out animator with android:interpolator="@android:interpolator/accelerate_cubic" and android:duration="1000". Then I made a splash theme. and connected both with transition in splashactivity.
> 
> So you made a new activity then and not the cold start method?

Yeah

**Comment by Bhavnaharitsa** [comment]
@devansh-299 Both the methods were same from the beginning. Now, that we've come to conclusion to implement the Activity method, do you want me to work on it?

**Comment by devansh-299** [comment]
> @devansh-299 Both the methods were same from the beginning. Now, that we've come to conclusion to implement the Activity method, do you want me to work on it?
> 

Yeah, please go ahead!

**Comment by devansh-299** [comment]
> @Bhavnaharitsa
> You don't need to create a separate layout file as loading the layout after initialization completely defies the the very use of Splash screens . You can read about "Cold Start" to know more about this .
> You can make a drawable file and pass it as arguement to android:windowBackground attribute .

@Bhavnaharitsa sorry if I couldn't explain properly, but you don't have to create a separate layout file for the SplashScreen. Just create a new activity and use that `SplashTheme` for that activity. I see that in this approach we can't add any animations but still, this approach is better as 
it handles the white screen appearing during the [ColdStart](https://developer.android.com/topic/performance/vitals/launch-time#cold)

**Comment by Bhavnaharitsa** [comment]
@devansh-299   when you implement cold start, the activity will be there but the layout XML file womt have any input but instead we create a drawable file and link it to a style and link that style to the activity in Manifest. Is that how it is?

**Comment by devansh-299** [comment]
> @devansh-299 when you implement cold start, the activity will be there but the layout XML file womt have any input but instead we create a drawable file and link it to a style and link that style to the activity in Manifest. Is that how it is?

Yes:)

**Comment by luckyman20** [comment]
@devansh-299 Sorry, I would not be able to participate furthermore in this PR. You can go ahead with your approach. It can later be tweaked anyways.

**Comment by rachittt** [comment]
hi @Bhavnaharitsa, kindly make the changes suggested by @devansh-299 in 2 days or else we will have to close this PR so that some other contributor can work on it because it has been under discussion for quite some time now; and we are trying to follow a schedule for certain features and bugs.

**Comment by Bhavnaharitsa** [comment]
@shiv07tiwari Please assign to some who can do it quickly. I hope @HaripriyaB has a solution. She might want to proceed with the pr:)
